addons/product_images/wizard/product_fetch_image_wizard.py

```diff
@@ -0,0 +1,238 @@
+            )
+        else:
+            if len(products_to_process) > 100:
+                self.env.ref('product_images.ir_cron_fetch_image')._trigger(
+                    fields.Datetime.now() + timedelta(minutes=1.0)
+                )
+
+    def _fetch_image_from_google(self, products):
+        """ Fetch images of a list of products from Google
+
+        Retrieve an image for each product in `products` from the Google Custom Search API
+        using their barcode. Fetch the 10 first images url and save the first valid image.
+
+        :param products: Product variant recordset
+        """
+        if not products:
+            return
+
+        ICP = self.env['ir.config_parameter']
+        session = requests.Session()
+        custom_search_req = requests.Request(
+            method='GET',
+            url='https://customsearch.googleapis.com/customsearch/v1',
+            params={
+                'cx': ICP.get_param('google.pse.id').strip(),
+                'safe': 'active',
+                'searchType': 'image',
+                'key': ICP.get_param('google.custom_search.key').strip(),
+                'rights': 'cc_publicdomain,cc_attribute,cc_sharealike',
+                'imgSize': 'large',
+                'imgType': 'photo',
+                'fields': 'searchInformation/totalResults,items(link)',
+            },
+        )
+        prepared_req = session.prepare_request(custom_search_req)
+        request_url = prepared_req.url
+
+        for product in products:
+            try:
+                prepared_req.prepare_url(request_url, {'q': product.barcode})
+
+                response = session.send(prepared_req, timeout=5)
+                response.raise_for_status()
+                response_content = response.json()
+            except (ConnectTimeout, ReadTimeout):
+                continue
+            except HTTPError as http_error:
+                if http_error.response.status_code == requests.codes.service_unavailable:
+                    continue
+                else:
+                    raise
+            else:
+                if int(response_content['searchInformation']['totalResults']) > 0:
+                    for item in response_content['items']:
+                        if 'x-raw-image:///' in item['link']:
+                            continue
+                        try:
+                            req = session.get(item['link'], timeout=5)
+                            req.raise_for_status()
+                            if 'image/' in req.headers['Content-Type']:
+                                product.image_1920 = base64.b64encode(req.content)
+                                break
+                        except Exception:
+                            continue
+            finally:
+                product.image_fetch_pending = False
+                # Commit every image to ensure that we don't lose a call
+                self.env.cr.commit()
```

Comment:
Why put this in the `finally` rather than in the `else`?

Reply:
Because the code in the else will not be triggered if there is an exception and, if any, I need to tag `image_fetch_pending` at` False` to avoid continuously making requests to Google for a product that doesn't have an image in their database.